src/GitHub.Api/Git/Repository.cs

```diff
@@ -669,27 +653,129 @@ public bool IsGitHub
             "{0} Owner: {1} Name: {2} CloneUrl: {3} LocalPath: {4} Branch: {5} Remote: {6}", GetHashCode(), Owner, Name,
             CloneUrl, LocalPath, CurrentBranch, CurrentRemote);
 
-        public IUser User { get; set; }
-
         protected static ILogging Logger { get; } = Logging.GetLogger<Repository>();
     }
 
     public interface IUser
     {
-        string Name { get; set; }
-        string Email { get; set; }
+        string Name { get; }
+        string Email { get; }
+        event Action<CacheUpdateEvent> Changed;
+        void CheckUserChangedEvent(CacheUpdateEvent cacheUpdateEvent);
+        void Initialize(IGitClient client);
+        void SetNameAndEmail(string name, string email);
     }
 
     [Serializable]
     public class User : IUser
     {
+        private ICacheContainer cacheContainer;
+        private IGitClient gitClient;
+
+        public event Action<CacheUpdateEvent> Changed;
+
+        public User(ICacheContainer cacheContainer)
+        {
+            this.cacheContainer = cacheContainer;
+
+            cacheContainer.GitUserCache.CacheInvalidated += GitUserCacheOnCacheInvalidated;
+            cacheContainer.GitUserCache.CacheUpdated += GitUserCacheOnCacheUpdated;
```

Comment:
Will the `User` objects never need to be GCd? Adding these event handlers will mean that `cacheContainer.GitUserCache` will retain a reference to each `User` object.

Reply:
`User` and `Repository` are found on `DefaultEnvironment`.

https://github.com/github-for-unity/Unity/blob/2e169082e46f436b395a5e53ccd2c2bed297251a/src/GitHub.Api/Platform/IEnvironment.cs#L30-L31

https://github.com/github-for-unity/Unity/blob/2e169082e46f436b395a5e53ccd2c2bed297251a/src/GitHub.Api/Platform/DefaultEnvironment.cs#L137-L138

There is generally only one instance of `User` and `Repository` in the system created at specific points of initialization.

https://github.com/github-for-unity/Unity/blob/2e169082e46f436b395a5e53ccd2c2bed297251a/src/GitHub.Api/Platform/DefaultEnvironment.cs#L51

https://github.com/github-for-unity/Unity/blob/2e169082e46f436b395a5e53ccd2c2bed297251a/src/GitHub.Api/Platform/DefaultEnvironment.cs#L89


These objects are only "destroyed" when the domain reloads. When the domain is reloaded these objects are being fed a `CacheContainer`. These objects store their data in file based caches specifically for this reason. It will take a moment on some other threads for the `User` and `Repository` object to get what it needs to be fully functioning (they both have `Initialize` methods for that). Until they are fully functioning, they get to use their file based cache to supply data to the user interface.